### PR TITLE
feat: PKCE magic link

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

This PR adds PKCE to  magic link and sign up verification types. The remaining flows - `email_change` and `recovery` will come soon. The `invite` type is subject to discussion.

We introduce a new column `authentication_method` to assis with tracking of  `amr` claims when we exchange an Auth Code for a session and  trigger the generation of an access token.

As an overview, the user will now receive a token `pkce_<token>` in the email when making a `/otp` request to PKCE - this will create a flow state containing the relevant details of the flow state. 

The user will then call `/verify` using  this token and GoTrue will search for the latest flow state for a user and attach a user's ID to that state. 

TODO: 
- [x] Write integration test
- [x] manually test against test app

### CODE REVIEW

File: internal/api/external.go [hunk 1]
```diff
@@ -390,7 +390,8 @@ func (a *API) createAccountFromExternalIdentity(tx *storage.Connection, r *http.
 		if !emailData.Verified && !config.Mailer.Autoconfirm {
 			mailer := a.Mailer(ctx)
 			referrer := a.getReferrer(r)
-			if terr = sendConfirmation(tx, user, mailer, config.SMTP.MaxFrequency, referrer, config.Mailer.OtpLength); terr != nil {
+			// TODO (joel) - add PKCE
+			if terr = sendConfirmation(tx, user, mailer, config.SMTP.MaxFrequency, referrer, config.Mailer.OtpLength, models.ImplicitFlow); terr != nil {
```
Comment (hf) [inline internal/api/external.go:394]: This todo supposed to be here?

File: internal/models/factor.go [hunk 2]
```diff
@@ -55,6 +55,22 @@ func (authMethod AuthenticationMethod) String() string {
...
+	case "password":
+		return PasswordGrant, nil
+	case "otp":
+		return OTP, nil
+	case "totp":
+		return TOTPSignIn, nil
+	case "sso/saml":
+		return SSOSAML, nil
+	}
+	return -1, errors.New("Invalid authentication method")
```
Comment (hf) [inline internal/models/factor.go:71]:
```suggestion
	return 0, fmt.Errorf("unsupported authentication method %q", authMethod)
```

File: internal/models/flow_state.go [hunk 3]
```diff
@@ -45,23 +47,61 @@ func (authMethod CodeChallengeMethod) String() string {
...
 }
 
+func ParseCodeChallengeMethod(authMethod string) (CodeChallengeMethod, error) {
+	switch strings.ToLower(authMethod) {
+	case "s256":
+		return SHA256, nil
+	case "plain":
+		return Plain, nil
+	}
+	return -1, errors.New("invalid code challenge method")
```
Comment (hf) [inline internal/models/flow_state.go:57]:
```suggestion
	return 0, fmt.Errorf("unsupported code_challenge method %q", authMethod)
```

File: internal/models/flow_state.go [hunk 4]
```diff
@@ -45,23 +47,61 @@ func (authMethod CodeChallengeMethod) String() string {
...
+}
+
+func ParseFlowType(flowType string) (FlowType, error) {
+	switch strings.ToLower(flowType) {
+	case "pkce":
+		return PKCEFlow, nil
+	case "implicit":
+		return ImplicitFlow, nil
+	}
+	return -1, errors.New("invalid flow type")
```
Comment (hf) [inline internal/models/flow_state.go:84]:
```suggestion
	return 0, fmt.Errorf("unsupported flow type %q", flowType)
```

File: openapi.yaml [hunk 5]
```diff
@@ -522,6 +546,13 @@ paths:
                   type: boolean
                 data:
                   type: object
+                code_challenge_method:
+                  type: enum
```
Comment (hf) [inline openapi.yaml:550]:
```suggestion
                  type: string
```

**Review by hf — COMMENTED** [review]
Looks good, would appreciate a glance over the suggestions before approving.

File: internal/api/external.go (hunk 1 above)
Comment (J0) [reply internal/api/external.go:394]: Yup, wanted to highlight the reason why it's using implicit flow and not making use of `flowType`

Comment (kangmingtay) [reply internal/api/external.go:394]: maybe we should just remove this block altogether:

```
if !emailData.Verified && !config.Mailer.Autoconfirm {
  ...
}
```

This check is basically for cases where the oauth email is not verified, then we send an email confirmation to the user. But maybe we should simply return an error here and ask the user to confirm their email with the oauth provider?

Comment (J0) [reply internal/api/external.go]: Removed the TODO for now, will open another PR for removing the check as it involves editing some tests

File: internal/api/helpers.go [hunk 6]
```diff
@@ -229,3 +229,10 @@ func isStringInSlice(checkValue string, list []string) bool {
 func getBodyBytes(req *http.Request) ([]byte, error) {
 	return utilities.GetBodyBytes(req)
 }
+
+func firstN(s string, n int) string {
+	if len(s) > n {
+		return s[:n]
+	}
+	return s
+}
```
Comment (hf) [inline internal/api/helpers.go:238]: Can't this be replaced with `s[:min(n, len(s))]` ?

File: internal/api/verify.go [hunk 7]
```diff
@@ -89,7 +89,14 @@ func (a *API) verifyGet(w http.ResponseWriter, r *http.Request) error {
...
 		err         error
 		token       *AccessTokenResponse
+		authCode    string
 	)
+	var flowType models.FlowType
+	if firstN(params.Token, 5) == PKCEPrefix {
+		flowType = models.PKCEFlow
+	} else {
+		flowType = models.ImplicitFlow
+	}
```
Comment (hf) [inline internal/api/verify.go:99]: TBH I'd rather this uses `strings.HasPrefix` instead.

```suggestion
	if strings.HasPrefix(params.Token, PKCEPrefix) {
		flowType = models.PKCEFlow
	} else {
		flowType = models.ImplicitFlow
	}
```

**Review by hf — APPROVED** [review]

File: internal/api/helpers.go (hunk 6 above)
Comment (hf) [reply internal/api/helpers.go:238]: Let's use `strings.HasPrefix` instead

**Pull request merged**